official/Mask_RCNN/mehdi_resnet_model.py

```diff
@@ -0,0 +1,574 @@
+      shape: The shape of the variable to get.
+      dtype: The dtype of the variable to get. Note that if this is a low
+        precision dtype, the variable will be created as a tf.float32 variable,
+        then cast to the appropriate dtype
+      *args: Additional arguments to pass unmodified to getter.
+      **kwargs: Additional keyword arguments to pass unmodified to getter.
+
+    Returns:
+      A variable which is cast to fp16 if necessary.
+    """
+
+    if dtype in CASTABLE_TYPES:
+      var = getter(name, shape, tf.float32, *args, **kwargs)
+      return tf.cast(var, dtype=dtype, name=name + '_cast')
+    else:
+      return getter(name, shape, dtype, *args, **kwargs)
+
+  def _model_variable_scope():
+    """Returns a variable scope that the model should be created under.
+
+    If self.dtype is a castable type, model variable will be created in fp32
+    then cast to self.dtype before being used.
+
+    Returns:
+      A variable scope for the model.
+    """
+
+    return tf.variable_scope('resnet_model',
+                             custom_getter=_custom_dtype_getter)
+
+  with _model_variable_scope():
+    if data_format == 'channels_first':
+      # Convert the inputs from channels_last (NHWC) to channels_first (NCHW).
+      # This provides a large performance boost on GPU. See
+      # https://www.tensorflow.org/performance/performance_guide#data_formats
+      inputs = tf.transpose(inputs, [0, 3, 1, 2])
+
+    inputs = conv2d_fixed_padding(
+        inputs=inputs, filters=num_filters, kernel_size=kernel_size,
+        strides=conv_stride, data_format=data_format)
+    inputs = tf.identity(inputs, 'initial_conv')
+
+    # We do not include batch normalization or activation functions in V2
+    # for the initial conv1 because the first ResNet unit will perform these
+    # for both the shortcut and non-shortcut paths as part of the first
+    # block's projection. Cf. Appendix of [2].
+    if resnet_version == 1:
+      inputs = batch_norm(inputs, training, data_format)
+      inputs = tf.nn.relu(inputs)
+
+    # feature pyramid
+    output = []
+    if first_pool_size:
+      inputs = tf.layers.max_pooling2d(
+          inputs=inputs, pool_size=first_pool_size,
+          strides=first_pool_stride, padding='SAME',
+          data_format=data_format)
+      inputs = tf.identity(inputs, 'initial_max_pool')
+      # add C1 to the feature pyramid
+      output.append(inputs)
+
+    for i, num_blocks in enumerate(block_sizes):
+      num_filters = num_filters * (2**i)
+      inputs = block_layer(
+          inputs=inputs, filters=num_filters, bottleneck=bottleneck,
+          block_fn=block_fn, blocks=num_blocks,
+          strides=block_strides[i], training=training,
+          name='block_layer{}'.format(i + 1), data_format=data_format)
+      # add C2,3,4,5 to the feature pyramid
+      output.append(inputs)
+
+    return output
+
+def fpn_graph(feature_pyramid, top_down_feature_dimension, data_format):
```

Comment:
Would be good to add a few comments inside this function explaining what's going on / giving a reference to the paper where the FPN is defined.